### PR TITLE
DM-46776: Add butler zip-from-graph command

### PR DESCRIPTION
**Pull request by timj**
Depends on lsst/daf_butler#1105

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: python/lsst/pipe/base/cli/cmd/commands.py [hunk 1]
```diff
@@ -62,10 +63,31 @@ def register_instrument(*args: Any, **kwargs: Any) -> None:
...
+@click.command(short_help="Make Zip archive from output files using graph.", cls=ButlerCommand)
+@click.argument("graph", required=True)
+@repo_argument(required=True)
+@click.argument("dest", required=True)
+@dataset_type_option(help="Dataset types to include in Zip archive.")
+@options_file_option()
+def zip_from_graph(**kwargs: Any) -> None:
+    """Transfer datasets from a quantum graph to a destination butler.
+
+    SOURCE is a URI to the source quantum graph file to use when building the
```
Comment (andy-slac) [inline python/lsst/pipe/base/cli/cmd/commands.py:84]: This should also be `GRAPH`.

File: python/lsst/pipe/base/cli/cmd/commands.py [hunk 2]
```diff
@@ -62,10 +63,31 @@ def register_instrument(*args: Any, **kwargs: Any) -> None:
...
+
+
+@click.command(short_help="Make Zip archive from output files using graph.", cls=ButlerCommand)
+@click.argument("graph", required=True)
+@repo_argument(required=True)
+@click.argument("dest", required=True)
+@dataset_type_option(help="Dataset types to include in Zip archive.")
+@options_file_option()
+def zip_from_graph(**kwargs: Any) -> None:
+    """Transfer datasets from a quantum graph to a destination butler.
```
Comment (andy-slac) [inline python/lsst/pipe/base/cli/cmd/commands.py:82]:
```suggestion
    """Transfer datasets from a quantum graph to a Zip archive.
```

File: python/lsst/pipe/base/cli/cmd/commands.py [hunk 3]
```diff
@@ -62,10 +63,31 @@ def register_instrument(*args: Any, **kwargs: Any) -> None:
...
+
+    SOURCE is a URI to the source quantum graph file to use when building the
+    Zip archive.
+
+    REPO is a URI to a butler configuration that is used to configure the
+    datastore of the quantum-backed butler.
+
+    DEST is a directory to write the Zip archive.
```
Comment (andy-slac) [inline python/lsst/pipe/base/cli/cmd/commands.py:90]: There is something very odd about help output, I suspect there is a big mess somewhere in MWArgumentDecorator. When I print usage of this command I get this:
```
$ butler zip-from-graph -h
Usage: butler zip-from-graph [OPTIONS] GRAPH REPO DEST

  Transfer datasets from a quantum graph to a destination butler.

  REPO is the URI or path to an existing data repository root or configuration
  file.

  SOURCE is a URI to the source quantum graph file to use when building the
  Zip archive.

  REPO is a URI to a butler configuration that is used to configure the
  datastore of the quantum-backed butler.

  DEST is a directory to write the Zip archive.

Options:
...
```
Note there are two mentions of REPO with different descriptions, and docstring only has one REPO string.

File: python/lsst/pipe/base/script/zip_from_graph.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    qgraph = QuantumGraph.loadUri(graph)
+
+    # Collect output refs that could be created by this graph.
+    original_output_refs: set[DatasetRef] = set(qgraph.globalInitOutputRefs())
+    for task_def in qgraph.iterTaskGraph():
+        if refs := qgraph.initOutputRefs(task_def):
+            original_output_refs.update(refs)
+    for qnode in qgraph:
+        for otherRefs in qnode.quantum.outputs.values():
+            original_output_refs.update(otherRefs)
```
Comment (andy-slac) [inline python/lsst/pipe/base/script/zip_from_graph.py:77]: Would it be useful to have that code as a method in the graph class?

**Review by andy-slac — APPROVED** [review]
Looks great.

File: python/lsst/pipe/base/cli/cmd/commands.py (hunk 3 above)
Comment (timj) [reply python/lsst/pipe/base/cli/cmd/commands.py:90]: Thanks. Looks like `repo_argument` brings its own help.

File: python/lsst/pipe/base/script/zip_from_graph.py (hunk 4 above)
Comment (timj) [reply python/lsst/pipe/base/script/zip_from_graph.py:77]: Probably.

**Pull request merged**